crates/wasmtime/src/runtime/rr/replay_driver.rs

```diff
@@ -0,0 +1,586 @@
+
+                    func.post_return(&mut store)?;
+                }
+                #[cfg(not(feature = "rr-component"))]
+                {
+                    let _ = event;
+                    bail!(
+                        "Cannot parse ComponentPostReturn replay event without rr-component feature enabled"
+                    );
+                }
+            }
+            RREvent::CoreWasmInstantiation(event) => {
+                // Find matching module from environment to instantiate
+                let module = self
+                    .env
+                    .modules
+                    .get(&event.module)
+                    .ok_or(ReplayError::MissingModule(event.module))?;
+
+                let instance = self
+                    .module_linker
+                    .instantiate(self.store.as_context_mut(), module)?;
+
+                // Validate the instantiation event
+                event.validate(&core_events::InstantiationEvent {
+                    module: *module.checksum(),
+                    instance: instance.id(),
+                })?;
+
+                self.module_instances.insert(instance.id(), instance);
+            }
+            RREvent::CoreWasmFuncEntry(event) => {
+                // Grab the correct module instance
+                let key = event.origin.instance;
+                let instance = self
+                    .module_instances
+                    .get_mut(&key)
+                    .ok_or(ReplayError::MissingModuleInstance(key.as_u32()))?;
+
+                let entity = EntityIndex::from(event.origin.index);
+                let mut store = self.store.as_context_mut();
+                let func = instance
+                    ._get_export(store.0, entity)
+                    .into_func()
+                    .ok_or(ReplayError::InvalidCoreFuncIndex(entity))?;
+
+                let params_ty = func.ty(&store).params().collect::<Vec<_>>();
+
+                // Obtain the argument values for function call
+                let mut results = vec![crate::Val::I64(0); func.ty(&store).results().len()];
+                let params = event.args.to_val_vec(&mut store, params_ty);
+
+                // Call the function
+                //
+                // This is almost a mirror of the usage in [`crate::Func::call_impl`]
+                func.call_impl_check_args(&mut store, &params, &mut results)?;
+                unsafe {
+                    func.call_impl_do_call(&mut store, params.as_slice(), results.as_mut_slice())?;
+                }
+            }
+
+            _ => {
+                log::error!("Unexpected top-level RR event: {:?}", rr_event);
+                Err(ReplayError::IncorrectEventVariant)?
+            }
+        }
+        Ok(())
+    }
+
+    /// Exactly like [`Self::run_single_top_level_event`] but uses async stores and calls
+    #[cfg(feature = "async")]
+    pub async fn run_single_top_level_event_async(&mut self, rr_event: RREvent) -> Result<()>
```

Comment:
This is a useful internal function (and non-async version above) but I'm not sure I see many useful ways to use it externally, once we have debugger support in general to single-step -- it exposes our internal notion of event divisions, which can change over time and is related to details of the component ABI (for example). I'd rather have "run the trace" be the single action one can take, as if the trace itself were one function invocation, and let other introspection/debugging stuff compose with that.